src/libstd/io/buffered.rs

```diff
@@ -588,7 +588,14 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => {
+                use crate::mem::forget;
+                use crate::ptr::{drop_in_place, read};
+                let inner = unsafe { read(&mut self.inner) };
```

Comment:
This approach of ~manually running the drop glue for all of the other parts of the struct seems a bit dangerous and error prone. Using `ManuallyDrop::take` seems like it'd be cleaner: https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html?search=#method.take

Reply:
What are you suggesting applying `ManuallyDrop::take` in this context?

Reply:
I think you would need the struct to contain:

```rust
    inner: ManuallyDrop<W>,
    buf: ManuallyDrop<Vec<u8>>,
```

where you call `ManuallyDrop::drop(&mut self.buf); ManuallyDrop::drop(&mut self.inner)` inside Drop and `ManuallyDrop::drop(&mut self.buf); ManuallyDrop::take(&mut self.inner)` inside into_inner.

Reply:
I don't like that implementation at all.  The entire point of this change is to isolate this edge case into one function rather than spreading it throughout the entire struct.

Reply:
Perhaps he meant putting `self` into a `ManuallyDrop` to prevent these double free issues?  That seems reasonable:

```
use std::mem::ManuallyDrop;
use std::ptr::{drop_in_place, read};
let mut s = ManuallyDrop::new(self);
let inner = unsafe { read(&mut s.inner) };
unsafe { drop_in_place(&mut s.buf) };
Ok(inner)
```